### PR TITLE
[CI] Bump to Go 1.21

### PR DESCRIPTION
**Pull request by shivanshu1333**
Bump to Go 1.21

Fixes https://github.com/open-telemetry/opentelemetry-collector-releases/issues/383

### CODE REVIEW

**Comment by mx-psi** [comment]
> @mx-psi is this good to go?

I think we should merge this after the v0.83.0 release cycle is completed. cc @djaglowski for that as release manager

**Review by codeboten — APPROVED** [review]

**Pull request merged**
